### ISSUE TITLE
[FR] Aggregated statistics for all threads in a benchmark

### ISSUE THREAD

**Issue by Tasemo** [google/benchmark#1604]
**Is your feature request related to a problem? Please describe.**
First Issue. I would like to report the speedup (T(1) / T(n)) when using multiple threads with ThreadRange(). There exists ComputeStatistics(), but it only contains values for one specific thread. It seems there is no way to get all values for a single named benchmark. Please correct me if I´m wrong.

**Describe the solution you'd like**
Ideally, there should be a method similiar to ComputeStatistics() that aggregates all values for all threads.

**Describe alternatives you've considered**
Maybe access to the State can be provided in ComputeStatistics(). This way, at least each run can be identified by its thread_index and its name. Values can then be stored externally.

**Comment by LebedevRI** [comment]
I'm not sure i fully understand the problem.
If the desired result is that the (same) custom user counter is accumulated from all threads,
then that is the default behavior:
https://github.com/google/benchmark/blob/9885aefb96effeb60c4e8c005e7b52c455458c10/test/user_counters_test.cc#L251-L285

**Comment by Tasemo** [comment]
Sorry, problem is not related to custom counters (at least I think). In the following code snipped, how can I report statistics for the whole "BM_Test" benchmark? It calls the lambda for each individual thread.
```c++
BENCHMARK(BM_Test)->ThreadRange(1, 8)->ComputeStatistics("speedup", [](const std::vector<double>& values) {
    // "values" are for one thread only, not the whole "BM_Test" benchmark
}
```

**Comment by LebedevRI** [comment]
Hm. Could you please try `MeasureProcessCPUTime()`?

**Comment by Tasemo** [comment]
Doesn´t work unfortunately. I think the problem is, that each thread gets treated as an individual benchmark internally. The context of which original benchmark they belong to is lost. The same applies to the Range*() functions and ArgsProduct().

**Comment by Tasemo** [comment]
For convenience, a project is available at [https://github.com/Tasemo/benchmark-issue-1604](https://github.com/Tasemo/benchmark-issue-1604).
<details>
  <summary>Code snippet</summary> 

```cpp  
#include <benchmark/benchmark.h>
#include <iostream>

static void BM_StringCopy(benchmark::State& state) {
  std::string x = "hello";
  for (auto _ : state) {
    std::string copy(x);
  }
}

double counter = 0.0;
BENCHMARK(BM_StringCopy)
  ->Repetitions(2)
  ->ThreadRange(1, 4)
  ->ComputeStatistics("test", [](const std::vector<double>& values) {
      std::cout << "Id: " << counter++ << ", Size: " << values.size() << std::endl;
      return 0.0;
    });
```

</details>

This code (linked with benchmark_main) prints the following pattern:
Id: 0, Size: 2
Id: 1, Size: 2
BM_StringCopy/repeats:2/threads:1 [...]
Id: 2, Size: 2
Id: 3, Size: 2
BM_StringCopy/repeats:2/threads:2 [...]
Id: 4, Size: 2
Id: 5, Size: 2
BM_StringCopy/repeats:2/threads:4 [...]

Each ComputeStatistics() lambda is called 2 times for each thread (I guess one for CPU time and one for wall time) with the run times for each repetition. I don´t know the best approach, but I would like to get the run times for all threads together so that I can report, for example, how much faster they get compared to using a single thread. One approach I can think of is a function that gets called with a std::vector<std::vector<double>> or any other 2D structure after all runs related to one benchmark have finished.

**Comment by LebedevRI** [comment]
By "together" would it be enough to get the sum of per-thread times?

**Comment by Tasemo** [comment]
No, I need the separate per-thread values in one place. I don´t see another way. The value of one run is dependent on the value of another. If this feature request is too big or too specific enough, let me know.
